src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+					}
+					else
+					{
+						EscapeAssemblyName(assemblySpec);
+					}
+
+					m_bHasAssemblySpec = true;
+				}
+
+			}
+			public void ToString(out string retString)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+				retString = m_pStr.ToString ();
+			}
+
+			public void Clear()
+			{
+				if (m_pStr != null)
+				{
+					m_pStr.Clear ();
+				}
+				m_bNestedName = false;
+				m_instNesting = 0;
+				m_bFirstInstArg = false;
+				m_parseState = ParseState.START;
+				m_bHasAssemblySpec = false;
+				m_bUseAngleBracketsForGenerics = false;
+				// m_stack.Clear (); looks like Mono's Stack type is missing Clear()
+				m_stack = new Stack ();
+			}
+
+			private void EscapeName (string name)
+			{
+				if (TypeString.ContainsReservedChar (name))
+				{
+					foreach (char c in name.AsSpan ())
+					{
+						if (c == '\0')
+							break;
+						if (TypeString.IsTypeNameReservedChar (c))
+							m_pStr.Append ('\\');
+						m_pStr.Append (c);
+					}
+				}
+				else
+					Append (name);
+			}
+
+			private void EscapeAssemblyName (string name)
+			{
+				Append (name);
+			}
+
+			private void EscapeEmbeddedAssemblyName (string name)
+			{
+				bool bContainsReservedChar = false;
+
+				foreach (char c in name.AsSpan ())
+				{
+					if (c == ']')
+					{
+						bContainsReservedChar = true;
+						break;
+					}
+				}
+
+				if (bContainsReservedChar)
+				{
+					foreach (char c in name.AsSpan ())
+					{
+						if (c == ']')
+							Append ('\\');
+
+						Append (c);
+					}
+				}
+				else
+				{
+					Append (name);
+				}
+
+			}
+
+			private bool CheckParseState (ParseState validState) { return (m_parseState & validState) != 0;}
+			private void Fail () { m_parseState = ParseState.ERROR; }
+			private void PushOpenGenericArgument ()
+			{
+				m_stack.Push (m_pStr.Length);
+			}
+			private void PopOpenGenericArgument ()
+			{
+				int index = (int) m_stack.Pop ();
+
+				if (!m_bHasAssemblySpec)
+					m_pStr.Remove (index - 1, 1);
+
+				m_bHasAssemblySpec = false;
+			}
+
+			// These were public in the original implementation
+			private void SetUseAngleBracketsForGenerics (bool value) { m_bUseAngleBracketsForGenerics = value; }
+			private void Append (string pStr)
+			{
+				foreach (char c in pStr.AsSpan ())
+				{
+					if (c == '\0')
+						break;
+					m_pStr.Append (c);
+				}
+			}
+
+			private void Append (char c) { m_pStr.Append (c); }
+
+		}
+
+		#region Static Members
+
+		// TypeNameBuilder is NOT thread safe NOR reliable
+		internal static string? ToString(Type type, Format format)
+		{
+			if (format == Format.FullName || format == Format.AssemblyQualifiedName)
+			{
+				if (!type.IsGenericTypeDefinition && type.ContainsGenericParameters)
+					return null;
+			}
+
+			TypeNameBuilder tnb = new TypeNameBuilder(new TypeNameBuilderImpl ());
```

Comment:
This construction does not make much sense to me. Would be better to change it to a struct and forget about the whole Clear/Dispose logic?